---
title: "Count_Data"
output: html_document
date: "2023-11-27"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# About

This script takes the counts matrices from `Formatting/data/*` and creates a Seurat object that can then be filtered based on the specifications of `Klann et al.` It first processes the raw gene counts matrix and creates a seurat object - then it processes the guide UMI count matrix, adding this as an assay to the seurat object

# Imports

```{r, message=FALSE}
library(Seurat)
library(ggplot2)
library(dplyr)
library(Matrix)
library(data.table)
```

# Read in Files

## Gene Count Data

Let's first import the necessary files, which can be found in this ENCODE project: [ENCSR464SPV](https://www.encodeproject.org/functional-characterization-experiments/ENCSR464SPV/)

```{r}
gene_data_file_path <- "data/Gene_Count_Data"
gene_matrix <- Read10X(gene_data_file_path)
```

## Guide Count Data

We can import the Guide Count Data from the following ENCODE project: [ENCFF938MGR](https://www.encodeproject.org/files/ENCFF938MGR/){.uri}

```{r}
guide_data_file_path <- "data/Guide_Count_Data"
guide_matrix <- Read10X(guide_data_file_path)
```

# Create Seurat Object

First let's create a Seurat object based off the gene counts data

```{r}
# Create the Seurat Object
seurat_object <- CreateSeuratObject(
  counts = gene_matrix,
  project = "RawCounts",
)
```

Now let's add the guide UMIs as an assay

```{r}
# Add guides as an assay
seurat_object[["GuideUMIs"]] <- CreateAssayObject(
  counts = guide_matrix
)
```

# Filter Seurat Object

-   Cell Filtering

    -   Cells with \>20% mitochondrial UMI counts were filtered out

    -   Cells with \< 10,000 transcript UMIs were filtered out

-   They don't filter the number of features (genes)?

    -   filtering the nCount_RNA takes care of this problem

```{r}
# Calculate the percent mitochondrial reads
seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^MT-")

# Create a violin plot of features
VlnPlot(seurat_object, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r}
# Filtering cells with more than 20% mitochondrial UMI counts
seurat_object <- subset(seurat_object, subset = percent.mt < 20)

# Filtering cells with less than 10,000 transcript UMIs
seurat_object <- subset(seurat_object, subset = nCount_RNA > 10000)
```

```{r}
# Create a violin plot of features
VlnPlot(seurat_object, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


# Guide Assignment

-   Guide Assignment

    -   Guides were assigned to cells based on two criteria:

        -   A gRNA has ≥ 5 UMI counts in the cell

        -   A gRNA represents ≥0.5% of the total UMI counts in the cell

```{r}
# Get the Guide UMI counts and store it as a dense matrix
temp <- as.matrix(seurat_object[["GuideUMIs"]]@counts)

# Get total UMI counts per cell
total_umi_per_cell <- colSums(temp)
```

```{r}
# Calculate conditions using vectorized operations
# gRNA has ≥ 5 UMI counts
condition1 <- temp >= 5

# gRNA represents ≥ 0.5% of total UMI counts per cell
condition2 <- (temp / total_umi_per_cell) >= 0.005
```

```{r}
# Create binary matrix based on conditions
perturb_status <- as.integer(condition1 & condition2)

# Reshape the perturb_status to match the dimensions of temp
perturb_status <- matrix(perturb_status, nrow = nrow(temp), ncol = ncol(temp), byrow = FALSE)

# Convert the reshaped matrix to a sparse matrix
perturb_status <- Matrix(perturb_status, sparse = TRUE)

# Add colnames and rownames
rownames(perturb_status) <- rownames(seurat_object[["GuideUMIs"]]@counts)
colnames(perturb_status) <- colnames(seurat_object[["GuideUMIs"]]@counts)

# Add the perturb_status matrix as a new assay
seurat_object[["perturb_status"]] <- CreateAssayObject(counts = perturb_status)
```

# Save Files for CRISPRi Pipeline

## Counts Matrix

```{r}
# Read the features data
features <- read.csv("data/Gene_Count_Data/features.tsv.gz", 
                         sep = "\t", quote = "", header = FALSE)

# Extract the count matrix from the Seurat object and prepare the data frame
data_matrix <- as.matrix(GetAssayData(object = seurat_object, layer = "counts"))

# Convert matrix to data frame and add row names as a separate column for gene symbols
df <- as.data.frame(data_matrix)
df$VECTOR <- rownames(data_matrix)

# Rearrange the columns to have VECTOR as the first column
df <- df[, c("VECTOR", colnames(data_matrix))]

# Replace the names of each gene symbol with its associated ENSEMBL ID
# Create a named vector for easy replacement
replace_vector <- setNames(features$V1, features$V2)

# Replace values in df$VECTOR based on the replace_vector
df$VECTOR <- replace_vector[df$VECTOR]

# Remove rows with NA values in VECTOR column (meaning they didn't match with ENSEMBL IDs)
df <- df[!is.na(df$VECTOR), ]

# Write the modified data frame to a file
dir.create("data/Final_CRISPRi_Inputs/", recursive = TRUE, showWarnings = FALSE)
write.table(df, 
            gzfile("data/Final_CRISPRi_Inputs/dge.txt.gz"), 
            sep = '\t', 
            row.names = FALSE, 
            col.names = TRUE, 
            quote = FALSE)
```

## Perturb Status

```{r}
# Extract matrix
data_matrix <- as.matrix(GetAssayData(object = seurat_object@assays$GuideUMIs, layer = "counts"))

# Convert matrix to data frame and add row names as "VECTOR" column without the spacer
df <- as.data.frame(data_matrix)

# Remove spacer names
df$VECTOR <- sub("-.*", "", rownames(data_matrix))

# Rearrange the columns to have VECTOR as the first column
df <- df[, c("VECTOR", colnames(data_matrix))]

# Save to file
write.table(df,
            gzfile("data/Final_CRISPRi_Inputs/perturb_status.txt.gz"),
            sep = '\t', 
            row.names = FALSE,
            col.names = TRUE, 
            quote = FALSE)
```